tests/filesystem.go

```diff
@@ -0,0 +1,517 @@
+	if err != nil {
+		t.Log(err)
+	}
+}
+
+func removeMountFolders(t *testing.T, partitions []*types.Partition) {
+	for _, p := range partitions {
+		err := os.RemoveAll(p.MountPath)
+		if err != nil {
+			t.Log(err)
+		}
+	}
+}
+
+// returns true if no error, false if error
+func runIgnition(t *testing.T, stage, root, configDir string, expectFail bool) bool {
+	args := []string{"-clear-cache", "-oem", "file", "-stage", stage, "-root", root}
+	cmd := exec.Command("ignition", args...)
+	t.Log("ignition", args)
+	cmd.Dir = configDir
+	out, err := cmd.CombinedOutput()
+	if err != nil && !expectFail {
+		t.Fatal(args, err, string(out))
+	}
+	return err == nil
+}
+
+// pickDevice will return the device corresponding to a partition with a given
+// label in the given imageFile
+func pickDevice(t *testing.T, partitions []*types.Partition, fileName string, label string) string {
+	number := -1
+	for _, p := range partitions {
+		if p.Label == label {
+			number = p.Number
+		}
+	}
+	if number == -1 {
+		return ""
+	}
+
+	args := []string{"-l", fileName}
+	kpartxOut := mustRun(t, "kpartx", args...)
+	re, err := regexp.Compile("/dev/(?P<device>[\\w\\d]+)")
+	if err != nil {
+		t.Fatal("compiling regexp:", err)
+	}
+	match := re.FindSubmatch(kpartxOut)
+	if len(match) < 2 {
+		t.Log(string(kpartxOut))
+		t.Fatal("couldn't find device")
+	}
+	return fmt.Sprintf("/dev/mapper/%sp%d", string(match[1]), number)
+}
+
+func writeIgnitionConfig(t *testing.T, config string) string {
+	tmpDir, err := ioutil.TempDir(os.Getenv("TMPDIR"), "config")
+	if err != nil {
+		t.Fatal(err)
+	}
+	err = ioutil.WriteFile(
+		filepath.Join(tmpDir, "config.ign"), []byte(config), 0644)
+	if err != nil {
+		t.Fatal(err)
+	}
+	return tmpDir
+}
+
+func calculateImageSize(partitions []*types.Partition) int64 {
+	// 63 is the number of sectors cgpt uses when generating a hybrid MBR
+	size := int64(63 * 512)
+	for _, p := range partitions {
+		size += int64(align(p.Length, 512) * 512)
+	}
+	size = size + int64(4096*512) // extra room to allow for alignments
+	return size
+}
+
+// createVolume will create the image file of the specified size, create a
+// partition table in it, and generate mount paths for every partition
+func createVolume(t *testing.T, imageFile string, size int64, cylinders int, heads int, sectorsPerTrack int, partitions []*types.Partition) {
+	// attempt to create the file, will leave already existing files alone.
+	// os.Truncate requires the file to already exist
+	out, err := os.Create(imageFile)
+	if err != nil {
+		t.Fatal("create", err, out)
+	}
+	out.Close()
+
+	// Truncate the file to the given size
+	err = os.Truncate(imageFile, size)
+	if err != nil {
+		t.Fatal("truncate", err)
+	}
+
+	createPartitionTable(t, imageFile, partitions)
+
+	for counter, partition := range partitions {
+		if partition.TypeCode == "blank" || partition.FilesystemType == "" {
+			continue
+		}
+
+		mntPath, err := ioutil.TempDir(os.Getenv("TMPDIR"), fmt.Sprintf("hd1p%d", counter))
```

Comment:
Currently we're creating one `TMPDIR` per test and then all the mount points and configs are placed as subdirectories inside of this. Should this not be the case?